### ISSUE TITLE
mapper issue

### ISSUE THREAD

**Issue by awcross1** [QISKit/qiskit-core#239]
The assertion at the bottom of yzy_to_zyz in _mapping.py can be triggered by some inputs because the threshold for "is zero", eps=1e-9, is too small. This is not a configurable parameter at the moment.

Here is the tail of the log output for one of these examples:

xi= 1.56685962124139
theta1= 0.784205150343885
theta2= 0.785398163397448
solutions= [(0.00411349627061388, 0.932520819437574, 0.638270810789719),
 (3.13747915731918, 0.932520819437574 + pi/2, 0.638270810789719 + pi/2),
 (0.00411349627061388, 0.932520819437574 + pi/2, -pi/2 + 0.638270810789719),
 (3.13747915731918, 0.932520819437574 + pi, 0.638270810789719)]
deltas= [5.79098781341297e-9,
 5.79098781861714e-9,
 0.00787337937646194,
 0.00787337937646193]

## Possible Solution

For the immediate future, we could increase eps, replace the assertion with an exception, and/or make the text of the exception more informative.
Eventually the passes should be configurable and control error.

## Your Environment

* Version used: master
* Environment name and version (e.g. Python 3.6.1): Python 3.6
* Operating System and version:

**Comment by lbello** [comment]
Which is the input?

**Comment by awcross1** [comment]
Here is an example that consistently produces the error.
[mapper_error.txt](https://github.com/QISKit/qiskit-sdk-py/files/1641638/mapper_error.txt)

**Comment by ajavadia** [comment]
@awcross1 @lbello 
I tried to localize this error. Made a tiny test case that produces the error:
```
faulty_qasm = """OPENQASM 2.0;
gate u3(theta,phi,lambda) q
{
  U(theta,phi,lambda) q;
}
gate u2(phi,lambda) q
{
  U((pi/2),phi,lambda) q;
}
gate u1(lambda) q
{
  U(0,0,lambda) q;
}
gate y a
{
  u3(pi,(pi/2),(pi/2)) a;
}
gate h a
{
  u2(0,pi) a;
}
gate s a
{
  u1((pi/2)) a;
}
qreg q[2];
creg c[2];
y q[0];
h q[0];
s q[0];
h q[0];
//U(pi,pi/2,pi/2) q[0];
//U(pi/2,0,pi) q[0];
//U(0,0,pi/2) q[0];
//U(pi/2,0,pi) q[0];
measure q[0] -> c[0];
measure q[1] -> c[1];"""

from qiskit import QuantumProgram
import logging
coupling_map = {}
coupling_map[0] = [1]
coupling_map[1] = []
qp = QuantumProgram()
qp.enable_logs(level=logging.DEBUG)
qc = qp.load_qasm_text(faulty_qasm, name="qc")
qobj = qp.compile([qc], coupling_map=coupling_map)
print("\nSUCCESS.")
```

Several things:
1. Strangely, the error goes away when I manually unroll the gates to U (see comments). There is something going on with extension gates.
2. Strangely, the error goes away when i reduce the size of the qubit register, even though the operations are entirely on q[0] and q[1] isn't doing anything.
3. When I relax the error threshold in `yzy_to_zyz` from `eps=1e-9` to `eps=1e-7`, the error in `_mapping.py` goes away, but I get precision errors from `qasm/_node/_real.py`

Still digging into it, but let me know if you have any leads.

**Comment by ajavadia** [comment]
Here are 2 tiny tests that trigger this error.
```python
qasm_string_1 = """OPENQASM 2.0;
include "qelib1.inc";
qreg qr[2];
cx qr[0],qr[1];
rz(0.7) qr[1];
rx(1.570796) qr[1];
"""
```
```python
qasm_string_2 = """OPENQASM 2.0;
include "qelib1.inc";
qreg qr[2];
y qr[0];
h qr[0];
s qr[0];
h qr[0];
"""
```
In both cases, the error occurs during compilation (specifically during `yzy_to_zyz`). To trigger it:
```python
from qiskit import *
import Qconfig
from pprint import pprint
register(Qconfig.APItoken)
circ = load_qasm_string(qasm_string_1)
qobj = compile(circ, 'ibmqx4')
pprint(qobj)
```

**Comment by jaygambetta** [comment]
This should be fixed by #607.

**Comment by ajavadia** [comment]
Closing, for the updated issue of #607. Moving some examples there.
